zebrad/src/commands/start/sync.rs

```diff
@@ -198,6 +198,24 @@ where
         while let Some(res) = requests.next().await {
             match res.map_err::<Report, _>(|e| eyre!(e)) {
                 Ok(zn::Response::BlockHashes(hashes)) => {
+                    tracing::trace!(?hashes);
+
+                    // zcashd sometimes appends an unrelated hash at the start
+                    // or end of its response.
+                    //
+                    // We can't discard the first hash, because it might be a
+                    // block we want to download. So we just accept any
+                    // out-of-order first hashes.
+
+                    // We use the last hash for the tip, and we want to avoid bad
+                    // tips. So we discard the last hash. (We don't need to worry
+                    // about missed downloads, because we will pick them up again
+                    // in ExtendTips.)
+                    let hashes = match hashes.split_last() {
+                        None => continue,
+                        Some((_last, rest)) => rest,
+                    };
```

Comment:
similarly, slice pattern might look nicer here, style choice tho, I'll leave that up to you.

Reply:
Refactored in 7ab0b0bf7745a5ced82529ec2d84f9af1d07c984